File: pipeline/internal/teamsuccess/teamsuccess.go
// Package teamsuccess calculates team and Mixed Draft player success ratings.
package teamsuccess

import (
	"fmt"
	"math"
	"sort"
	"strconv"
	"time"

	"github.com/aegis-draft/pipeline/internal/model"
	"github.com/aegis-draft/pipeline/internal/rating"
)

type EventTier string

const (
	TierTI    EventTier = "ti"
	TierMajor EventTier = "major"
	TierOne   EventTier = "tier1"
)

type TeamEventResult struct {
	EventID   string
	TeamID    int
	EndTime   int64
	Tier      EventTier
	Placement int
	PrizeUSD  int
}

type TeamMatchResult struct {
	MatchID   int64
	TeamID    int
	StartTime int64
	Won       bool
}

type PlayerInput struct {
	AccountID     int
	IndividualOVR int
	Teams         []PlayerTeamGames
}

type PlayerTeamGames struct {
	TeamID int
	Games  int
}

type PlayerScore struct {
	AccountID     int
	Games         int
	TeamScore     float64
	IndividualOVR int
	Score         int
}

type accumulator struct {
	placementPoints float64
	titles, top4    int
	prize, games    int
	wins            int
	tiPlacement     int
}

var timeWindows = []struct {
	format model.Format
	years  int
}{
	{model.Last1y, 1},
	{model.Last2y, 2},
	{model.Last5y, 5},
}

// Calculate builds teamSuccess.json-compatible metrics for last_1y/2y/5y.
// Valve Legacy remains a separately curated event set.
func Calculate(asOf time.Time, events []TeamEventResult, matches []TeamMatchResult, cfg rating.Config) (map[string]map[model.Format]model.TeamWindowSuccess, error) {
	if asOf.IsZero() {
		return nil, fmt.Errorf("team success as-of date is required")
	}
	if err := validateConfig(cfg); err != nil {
		return nil, err
	}
	if err := validateInputs(events, matches); err != nil {
		return nil, err
	}
	asOfDate := utcDate(asOf)
	result := make(map[string]map[model.Format]model.TeamWindowSuccess)
	for _, window := range timeWindows {
		start := asOfDate.AddDate(-window.years, 0, 0)
		teams := make(map[int]*accumulator)
		for _, event := range events {
			date := utcDate(time.Unix(event.EndTime, 0))
			if date.Before(start) || date.After(asOfDate) {
				continue
			}
			entry := ensureAccumulator(teams, event.TeamID)
			entry.placementPoints += prestige(event.Tier, cfg) * placementValue(event.Placement, cfg)
			if entry.prize > maxInt()-event.PrizeUSD {
				return nil, fmt.Errorf("team %d prize total overflows int", event.TeamID)
			}
			entry.prize += event.PrizeUSD
			if event.Placement == 1 {
				entry.titles++
			}
			if event.Placement <= 4 {
				entry.top4++
			}
			if event.Tier == TierTI && (entry.tiPlacement == 0 || event.Placement < entry.tiPlacement) {
				entry.tiPlacement = event.Placement
			}
		}
		for _, match := range matches {
			date := utcDate(time.Unix(match.StartTime, 0))
			if date.Before(start) || date.After(asOfDate) {
				continue
			}
			entry := ensureAccumulator(teams, match.TeamID)
			entry.games++
			if match.Won {
				entry.wins++
			}
		}
		teamIDs := make([]int, 0, len(teams))
		for teamID := range teams {
			teamIDs = append(teamIDs, teamID)
		}
		sort.Ints(teamIDs)
		for _, teamID := range teamIDs {
			key := strconv.Itoa(teamID)
			if result[key] == nil {
				result[key] = make(map[model.Format]model.TeamWindowSuccess)
			}
			result[key][window.format] = scoreTeam(teams[teamID], cfg)
		}
	}
	return result, nil
}

// ScorePlayers uses team success as the main signal and individual OVR as a
// bounded correction. Multi-team careers are weighted by games.
func ScorePlayers(format model.Format, teams map[string]map[model.Format]model.TeamWindowSuccess, players []PlayerInput, cfg rating.Config) ([]PlayerScore, error) {
	if format != model.Last1y && format != model.Last2y && format != model.Last5y {
		return nil, fmt.Errorf("unsupported team-success format %q", format)
	}
	if err := validateConfig(cfg); err != nil {
		return nil, err
	}
	result := make([]PlayerScore, 0, len(players))
	seenPlayers := make(map[int]struct{}, len(players))
	for _, player := range players {
		if player.AccountID <= 0 || player.IndividualOVR < 0 || player.IndividualOVR > 100 || len(player.Teams) == 0 {
			return nil, fmt.Errorf("invalid team-success player input: %+v", player)
		}
		if _, exists := seenPlayers[player.AccountID]; exists {
			return nil, fmt.Errorf("duplicate team-success player %d", player.AccountID)
		}
		seenPlayers[player.AccountID] = struct{}{}
		seenTeams := make(map[int]struct{}, len(player.Teams))
		weighted, games := 0.0, 0
		for _, stint := range player.Teams {
			if stint.TeamID <= 0 || stint.Games <= 0 {
				return nil, fmt.Errorf("invalid team stint for player %d: %+v", player.AccountID, stint)
			}
			if _, exists := seenTeams[stint.TeamID]; exists {
				return nil, fmt.Errorf("duplicate team %d for player %d", stint.TeamID, player.AccountID)
			}
			seenTeams[stint.TeamID] = struct{}{}
			windows, exists := teams[strconv.Itoa(stint.TeamID)]
			metric, existsForWindow := windows[format]
			if !exists || !existsForWindow {
				return nil, fmt.Errorf("missing team success for team %d format %s", stint.TeamID, format)
			}
			if !finite(metric.SuccessScore) || metric.SuccessScore < 0 || metric.SuccessScore > 100 {
				return nil, fmt.Errorf("invalid success score for team %d format %s", stint.TeamID, format)
			}
			if games > maxInt()-stint.Games {
				return nil, fmt.Errorf("player %d games overflow int", player.AccountID)
			}
			weighted += metric.SuccessScore * float64(stint.Games)
			games += stint.Games
		}
		teamScore := weighted / float64(games)
		factor := cfg.PlayerFormMinFactor + (cfg.PlayerFormMaxFactor-cfg.PlayerFormMinFactor)*float64(player.IndividualOVR)/100
		result = append(result, PlayerScore{
			AccountID: player.AccountID, Games: games, TeamScore: round2(teamScore),
			IndividualOVR: player.IndividualOVR, Score: rounded(teamScore * factor),
		})
	}
	sort.Slice(result, func(i, j int) bool { return result[i].AccountID < result[j].AccountID })
	return result, nil
}

func scoreTeam(entry *accumulator, cfg rating.Config) model.TeamWindowSuccess {
	placementScore := clamp100(entry.placementPoints * cfg.PlacementPointScale)
	prizeScore := 0.0
	if entry.prize > 0 {
		prizeScore = clamp100(100 * math.Log1p(float64(entry.prize)) / math.Log1p(cfg.PrizeReferenceUSD))
	}
	smoothedWinrate := (float64(entry.wins) + cfg.SmoothM*cfg.SmoothMu) / (float64(entry.games) + cfg.SmoothM)
	topScore := clamp100(float64(entry.titles)*cfg.TitlePoints + float64(entry.top4)*cfg.TopFinishPoints)
	score := placementScore*cfg.TeamSuccessWeights.Placement + prizeScore*cfg.TeamSuccessWeights.Prize +
		smoothedWinrate*100*cfg.TeamSuccessWeights.Winrate + topScore*cfg.TeamSuccessWeights.TopFinish
	rawWinrate := 0.0
	if entry.games > 0 {
		rawWinrate = float64(entry.wins) / float64(entry.games)
	}
	return model.TeamWindowSuccess{
		SuccessScore: round2(score), Titles: entry.titles, TopFinishes: entry.top4,
		PrizeUsd: entry.prize, Games: entry.games, Winrate: rawWinrate, TIPlacement: entry.tiPlacement,
	}
}

func validateInputs(events []TeamEventResult, matches []TeamMatchResult) error {
	type eventTeamKey struct {
		eventID string
		teamID  int
	}
	seenEvents := make(map[eventTeamKey]struct{}, len(events))
	for _, event := range events {
		if event.EventID == "" || event.TeamID <= 0 || event.EndTime <= 0 || event.Placement <= 0 || event.PrizeUSD < 0 || !validTier(event.Tier) {
			return fmt.Errorf("invalid team event result: %+v", event)
		}
		key := eventTeamKey{eventID: event.EventID, teamID: event.TeamID}
		if _, exists := seenEvents[key]; exists {
			return fmt.Errorf("duplicate event/team result %s/%d", event.EventID, event.TeamID)
		}
		seenEvents[key] = struct{}{}
	}
	seenMatches := make(map[[2]int64]struct{}, len(matches))
	for _, match := range matches {
		if match.MatchID <= 0 || match.TeamID <= 0 || match.StartTime <= 0 {
			return fmt.Errorf("invalid team match result: %+v", match)
		}
		key := [2]int64{match.MatchID, int64(match.TeamID)}
		if _, exists := seenMatches[key]; exists {
			return fmt.Errorf("duplicate match/team result %d/%d", match.MatchID, match.TeamID)
		}
		seenMatches[key] = struct{}{}
	}
	return nil
}

func validateConfig(cfg rating.Config) error {
	weights := cfg.TeamSuccessWeights
	if !weightsSumToOne(weights.Placement, weights.Prize, weights.Winrate, weights.TopFinish) ||
		!finite(cfg.SmoothMu) || !finite(cfg.SmoothM) || cfg.SmoothMu < 0 || cfg.SmoothMu > 1 || cfg.SmoothM <= 0 ||
		!finite(cfg.EventPrestige.TI) || !finite(cfg.EventPrestige.Major) || !finite(cfg.EventPrestige.Tier1) ||
		cfg.EventPrestige.TI <= cfg.EventPrestige.Major || cfg.EventPrestige.Major <= cfg.EventPrestige.Tier1 || cfg.EventPrestige.Tier1 <= 0 ||
		!finite(cfg.PlacementWeights.Champion) || !finite(cfg.PlacementWeights.RunnerUp) || !finite(cfg.PlacementWeights.Top4) || !finite(cfg.PlacementWeights.Top8) ||
		cfg.PlacementWeights.Champion < cfg.PlacementWeights.RunnerUp || cfg.PlacementWeights.RunnerUp < cfg.PlacementWeights.Top4 ||
		cfg.PlacementWeights.Top4 < cfg.PlacementWeights.Top8 || cfg.PlacementWeights.Top8 < 0 ||
		!finite(cfg.PlacementPointScale) || !finite(cfg.PrizeReferenceUSD) || !finite(cfg.TitlePoints) || !finite(cfg.TopFinishPoints) ||
		cfg.PlacementPointScale <= 0 || cfg.PrizeReferenceUSD <= 0 || cfg.TitlePoints < 0 || cfg.TopFinishPoints < 0 ||
		!finite(cfg.PlayerFormMinFactor) || !finite(cfg.PlayerFormMaxFactor) ||
		cfg.PlayerFormMinFactor < 0 || cfg.PlayerFormMaxFactor < cfg.PlayerFormMinFactor ||
		math.Abs((cfg.PlayerFormMinFactor+cfg.PlayerFormMaxFactor)/2-1) > 1e-9 {
		return fmt.Errorf("invalid team-success rating config")
	}
	return nil
}

func prestige(tier EventTier, cfg rating.Config) float64 {
	switch tier {
	case TierTI:
		return cfg.EventPrestige.TI
	case TierMajor:
		return cfg.EventPrestige.Major
	default:
		return cfg.EventPrestige.Tier1
	}
}

func placementValue(placement int, cfg rating.Config) float64 {
	switch {
	case placement == 1:
		return cfg.PlacementWeights.Champion
	case placement == 2:
		return cfg.PlacementWeights.RunnerUp
	case placement <= 4:
		return cfg.PlacementWeights.Top4
	case placement <= 8:
		return cfg.PlacementWeights.Top8
	default:
		return 0
	}
}

func validTier(tier EventTier) bool { return tier == TierTI || tier == TierMajor || tier == TierOne }

func ensureAccumulator(teams map[int]*accumulator, teamID int) *accumulator {
	if teams[teamID] == nil {
		teams[teamID] = &accumulator{}
	}
	return teams[teamID]
}

func weightsSumToOne(weights ...float64) bool {
	sum := 0.0
	for _, weight := range weights {
		if math.IsNaN(weight) || math.IsInf(weight, 0) || weight < 0 {
			return false
		}
		sum += weight
	}
	return math.Abs(sum-1) <= 1e-9
}

func finite(value float64) bool { return !math.IsNaN(value) && !math.IsInf(value, 0) }

func clamp100(value float64) float64 { return math.Max(0, math.Min(100, value)) }
func rounded(value float64) int      { return int(math.Round(clamp100(value))) }
func round2(value float64) float64   { return math.Round(value*100) / 100 }
func maxInt() int                    { return int(^uint(0) >> 1) }

func utcDate(value time.Time) time.Time {
	utc := value.UTC()
	return time.Date(utc.Year(), utc.Month(), utc.Day(), 0, 0, 0, 0, time.UTC)
}

File: web/src/game/teamSuccess.ts
// Mixed Draft: base игрока = успех его команды за окно, а не форма на конкретном событии
// (PRD §5.3, §5.4.3). Скилл scoring-model, раздел «Base».
//
// ЗАЧЕМ. В Mixed пятёрка собрана из разных команд, общего события у них нет, поэтому event
// OVR там несопоставим: замер по packs.json — один и тот же игрок встречается в 62 паках с
// OVR 60..91 (разброс 31). То есть число говорит не про игрока, а про то, какой ивент выпал
// рандому. Успех команды за окно — свойство игрока в этом окне, и оно сравнимо между командами.
//
// ЧТО СЕЙЧАС В ДАННЫХ (важно не переоценивать). `successScore` приходит из прокси-агрегатора
// pipeline/internal/domain.BuildTeamSuccess = сглаженный winrate × вес тира лиги. Компоненты
// PRD «плейсменты 40% + призовые 20% + топ-финиши 15%» в данных НУЛЕВЫЕ (проверено: 417 записей
// окон, titles/topFinishes/prizeUsd = 0 во всех) — они выводятся из Liquipedia, а T1.3 ⛔.
// Полная реализация v1.2.0 лежит в pipeline/internal/teamsuccess, но не подключена.
// Когда Liquipedia приедет, поменяются ВХОДНЫЕ данные, а нормализация ниже останется той же.
//
// ПОЧЕМУ НОРМАЛИЗАЦИЯ ЗДЕСЬ, А НЕ В GO. Сырой successScore живёт в своей шкале (замер:
// медиана 32, диапазон 15.6..54.7), а игровая шкала — event OVR (медиана 74, 54..99) и сила
// ботов Normal(86,5). Подставить сырое число в base нельзя: команда получила бы Team OVR ~38
// против ботов ~86 и по ELO проигрывала бы каждую серию с вероятностью ~99%. Отображение
// одной шкалы на другую — это баланс, как и сам `Team OVR = base + synergy + chemistry`,
// поэтому оно живёт на фронте и меняется без пересборки датасета (см. data-contract,
// инвариант «сырые числа в данных, сглаживание/калибровка на клиенте»).

import type { Format, PackPlayer, TeamSuccess } from "../types/data.ts";

/** Параметры отображения успеха команды в игровую шкалу. Версионируются с моделью. */
export const MIXED_BASE = {
  /** Полоса силы команды ДО поправки на игрока. Калибрована по РЕАЛИЗОВАННОЙ медиане Team OVR
   *  (60 забегов на стиль, реальный датасет), а не по медиане когорты: игрок берёт не среднюю
   *  команду, а лучшую доступную, поэтому «середина полосы = медиана OVR» давала перекос
   *  Mixed +5 очков (83.4 против 78.3) и делала режим строго выгоднее. Замер на 48..76:
   *  Team Packs 78.3 (p10 72.6, p90 84.2) · Mixed 76.8 (p10 70.0, p90 83.8) — распределения
   *  перекрываются, ни один стиль не доминирует, оба играбельны против ботов Normal(86,5).
   *  Потолок 76 держит 76 × 1.2 = 91.2 ниже клампа 100: на 95 верхний перцентиль насыщался и
   *  терял разрешение (сильная и очень сильная команда давали одно число) — поймано тестом.
   *  Данные поменяются существенно → перепроверить тем же замером. */
  min: 48,
  max: 76,
  /** Поправка на индивидуальную форму, дословно PRD §5.4.3: 0.8 + 0.4·OVR/100 (OVR 50 нейтрален). */
  factorBase: 0.8,
  factorSpan: 0.4,
} as const;

/** Когорта одного format-scope: отсортированные successScore всех команд с данными.
 *  Скилл требует явный scope — окна не должны смешиваться молча. */
export type SuccessCohort = { format: Format; sorted: number[] };

export function buildSuccessCohort(teamSuccess: TeamSuccess, format: Format): SuccessCohort {
  const sorted: number[] = [];
  for (const windows of Object.values(teamSuccess)) {
    const score = windows[format]?.successScore;
    if (typeof score === "number" && Number.isFinite(score)) sorted.push(score);
  }
  sorted.sort((a, b) => a - b);
  return { format, sorted };
}

/** Есть ли у команды успех за это окно. Mixed-пак не должен предлагать команду без данных:
 *  PRD запрещает нейтральный fallback, а бросать исключение посреди забега — хуже, чем
 *  не показать команду вовсе (см. фильтр в generatePack). */
export function hasTeamSuccess(teamSuccess: TeamSuccess, teamId: number, format: Format): boolean {
  const score = teamSuccess[String(teamId)]?.[format]?.successScore;
  return typeof score === "number" && Number.isFinite(score);
}

/** Mixed-паку нужны 5 игроков из 5 РАЗНЫХ команд, поэтому меньше пяти команд с данными
 *  в окне — режим в нём неиграбелен. */
export const MIXED_MIN_TEAMS = 5;

/** Доступен ли Mixed в этом окне. Проверяем по данным, а не по имени формата: сегодня пустой
 *  только valve_legacy (плейсменты/призовые ждут Liquipedia, T1.3 ⛔), и когда он наполнится,
 *  режим откроется сам — без правки кода. */
export function mixedSupportsFormat(teamSuccess: TeamSuccess, format: Format): boolean {
  return buildSuccessCohort(teamSuccess, format).sorted.length >= MIXED_MIN_TEAMS;
}

/** Перцентиль значения в когорте → [0,1]. Пустая когорта → 0.5 (нейтраль). */
function percentile(cohort: SuccessCohort, score: number): number {
  const n = cohort.sorted.length;
  if (n === 0) return 0.5;
  if (n === 1) return 0.5;
  // Доля команд строго слабее + половина равных — устойчиво к дубликатам скора.
  let below = 0;
  let equal = 0;
  for (const value of cohort.sorted) {
    if (value < score) below += 1;
    else if (value === score) equal += 1;
  }
  return (below + equal / 2) / n;
}

/** Успех команды в игровой шкале (без поправки на игрока). */
export function teamStrength(cohort: SuccessCohort, teamSuccess: TeamSuccess, teamId: number): number {
  const score = teamSuccess[String(teamId)]?.[cohort.format]?.successScore;
  if (typeof score !== "number" || !Number.isFinite(score)) {
    // Сюда не должны доходить: generatePack отфильтровывает такие команды заранее.
    throw new Error(`Нет team-success для команды ${teamId} в окне ${cohort.format}`);
  }
  const p = percentile(cohort, score);
  return MIXED_BASE.min + p * (MIXED_BASE.max - MIXED_BASE.min);
}

/** Base одного игрока в Mixed: успех команды × ограниченная поправка на его форму. */
export function mixedPlayerBase(
  cohort: SuccessCohort,
  teamSuccess: TeamSuccess,
  teamId: number,
  ovr: number,
): number {
  const team = teamStrength(cohort, teamSuccess, teamId);
  const factor = MIXED_BASE.factorBase + MIXED_BASE.factorSpan * (ovr / 100);
  return Math.min(100, Math.max(0, team * factor));
}

/** Base пятёрки в Mixed — среднее по игрокам, как и event-версия (baseRating). */
export function mixedBaseRating(
  players: PackPlayer[],
  teamIdByPlayer: Map<number, number>,
  teamSuccess: TeamSuccess,
  format: Format,
): number {
  if (players.length === 0) return 0;
  const cohort = buildSuccessCohort(teamSuccess, format);
  let sum = 0;
  for (const player of players) {
    const teamId = teamIdByPlayer.get(player.accountId);
    if (teamId === undefined) throw new Error(`Нет teamId для игрока ${player.accountId} в Mixed`);
    sum += mixedPlayerBase(cohort, teamSuccess, teamId, player.ovr);
  }
  return sum / players.length;
}
